src/Kestrel.Transport.Sockets/Internal/SocketConnection.cs

```diff
@@ -244,7 +244,14 @@ public void Dispose()
 
                 // This is not interlocked because there could be a concurrent writer.
                 // Instead it's to prevent read tearing on 32-bit systems.
-                Interlocked.Add(ref _totalBytesWritten, buffer.Length);
+                if (IntPtr.Size == sizeof(long))
+                {
+                    _totalBytesWritten += buffer.Length;
```

Comment:
@halter73 my comment/question is about removing `Interlocked` here.
Reads and writes have semantics about how observable they are by other threads. Maybe by removing the `Interlocked` here, the change to `_totalBytesWritten` can go unnoticed by other threads? Or perhaps the C#/.NET memory model requires such writes (e.g. to member fields) to always be visible? I'm not sure.
In my own code I play safe and when using `Volatile.Read` for reading, I use it for writing too.

Reply:
Good point.

> the C#/.NET memory model requires such writes (e.g. to member fields) to always be visible?

To my understanding this guarantee isn't given by the ECMA standard.
x86/amd64 usually have a strong memory model, so it is fine (and I just had this architecture in mind when doing the PR).
arm has a weak memory model, so it must not be fine, and your concern seems correct.

To be on the safe side we should use https://github.com/aspnet/KestrelHttpServer/pull/2593#discussion_r189871592 for both 32/64 bit? Or fallback to `Interlocked` but it's 4x slower than the `Volatile`-variant.

Reply:
I think we can use either 'Volatile' or 'Interlocked' and that will be correct on 32-bit and 64-bit. So, we can pick the fastest and have the same code for both architectures.